File: Bridge/System/Double.cs
using Bridge;

namespace System
{
    [External]
    [Name("Bridge.Double")]
    [Constructor("Number")]
    public struct Double : IComparable, IComparable<Double>, IEquatable<Double>, IFormattable
    {
        private extern Double(int i);

        [Name("max")]
        public const double MaxValue = 1.7976931348623157E+308;

        [Name("min")]
        public const double MinValue = 1.7976931348623157E+308;

        [InlineConst]
        public const double Epsilon = 4.94065645841247E-324;

        [Template("Number.NEGATIVE_INFINITY")]
        public const double NegativeInfinity = -1.0/0.0;

        [Template("Number.POSITIVE_INFINITY")]
        public const double PositiveInfinity = 1.0/0.0;

        [Template("Number.NaN")]
        public const double NaN = 0.0/0.0;

        [Template("Bridge.Double.format({this}, {format})")]
        public extern string Format(string format);

        [Template("Bridge.Double.format({this}, {format}, {provider})")]
        public extern string Format(string format, IFormatProvider provider);

        public extern string ToString(int radix);

        [Template("Bridge.Double.format({this}, {format})")]
        public extern string ToString(string format);

        [Template("Bridge.Double.format({this}, {format}, {provider})")]
        public extern string ToString(string format, IFormatProvider provider);

        [Template("Bridge.Double.format({this}, 'G')")]
        public override extern string ToString();

        [Template("Bridge.Double.format({this}, 'G', {provider})")]
        public extern string ToString(IFormatProvider provider);

        [Template("Bridge.Double.parse({s})")]
        public static extern double Parse(string s);

        [Template("Bridge.Int.parseFloat({s}, {provider})")]
        public static extern double Parse(string s, IFormatProvider provider);

        [Template("Bridge.Double.tryParse({s}, null, {result})")]
        public static extern bool TryParse(string s, out double result);

        [Template("Bridge.Double.tryParse({s}, {provider}, {result})")]
        public static extern bool TryParse(string s, IFormatProvider provider, out double result);

        public extern string ToExponential();

        public extern string ToExponential(int fractionDigits);

        public extern string ToFixed();

        public extern string ToFixed(int fractionDigits);

        public extern string ToPrecision();

        public extern string ToPrecision(int precision);

        [Template("({d} === Number.POSITIVE_INFINITY)")]
        public static extern bool IsPositiveInfinity(double d);

        [Template("({d} === Number.NEGATIVE_INFINITY)")]
        public static extern bool IsNegativeInfinity(double d);

        [Template("(Math.abs({d}) === Number.POSITIVE_INFINITY)")]
        public static extern bool IsInfinity(double d);

        [Template("isFinite({d})")]
        public static extern bool IsFinite(double d);

        [Template("isNaN({d})")]
        public static extern bool IsNaN(double d);

        [Template("Bridge.compare({this}, {other})")]
        public extern int CompareTo(double other);

        [Template("Bridge.compare({this}, {obj})")]
        public extern int CompareTo(object obj);

        [Template("{this} === {other}")]
        public extern bool Equals(double other);
    }
}

File: Bridge/Resources/Double.js
    Bridge.define("System.Double", {
        inherits: [System.IComparable, System.IFormattable],
        statics: {
            min: -Number.MAX_VALUE,

            max: Number.MAX_VALUE,

            precision: 15,

            $number: true,

            $is: function (instance) {
                return typeof (instance) === "number";
            },

            getDefaultValue: function () {
                return 0;
            },

            parse: function (s, provider) {
                return Bridge.Int.parseFloat(s, provider);
            },

            tryParse: function (s, provider, result) {
                return Bridge.Int.tryParseFloat(s, provider, result);
            },

            format: function (number, format, provider) {
                return Bridge.Int.format(number, format, provider, System.Double);
            },

            equals: function (v1, v2) {
                if (Bridge.is(v1, System.Double) && Bridge.is(v2, System.Double)) {
                    return Bridge.unbox(v1) === Bridge.unbox(v2);
                }

                return false;
            },

            equalsT: function (v1, v2) {
                return Bridge.unbox(v1) === Bridge.unbox(v2);
            },

            getHashCode: function (v) {
                var value = Bridge.unbox(v);
                if (value === Number.POSITIVE_INFINITY) {
                    return 0x7FF00000;
                }

                if (value === Number.NEGATIVE_INFINITY) {
                    return 0xFFF00000;
                }
                return value.toExponential();
            }
        }
    });

    System.Double.$kind = "";
    Bridge.Class.addExtend(System.Double, [System.IComparable$1(System.Double), System.IEquatable$1(System.Double)]);

    Bridge.define("System.Single", {
        inherits: [System.IComparable, System.IFormattable],
        statics: {
            min: -3.40282346638528859e+38,

            max: 3.40282346638528859e+38,

            precision: 7,

            $number: true,

            $is: System.Double.$is,

            getDefaultValue: System.Double.getDefaultValue,

            parse: System.Double.parse,

            tryParse: System.Double.tryParse,

            format: function (number, format, provider) {
                return Bridge.Int.format(number, format, provider, System.Single);
            },

            equals: function (v1, v2) {
                if (Bridge.is(v1, System.Single) && Bridge.is(v2, System.Single)) {
                    return Bridge.unbox(v1) === Bridge.unbox(v2);
                }

                return false;
            },

            equalsT: function (v1, v2) {
                return Bridge.unbox(v1) === Bridge.unbox(v2);
            },

            getHashCode: System.Double.getHashCode
        }
    });

    System.Single.$kind = "";
    Bridge.Class.addExtend(System.Single, [System.IComparable$1(System.Single), System.IEquatable$1(System.Single)]);
